File: src/io/github/hulang1024/chinesechess/rule/chess/ChessG.java
package io.github.hulang1024.chinesechess.rule.chess;

import io.github.hulang1024.chinesechess.rule.AbstractChess;
import io.github.hulang1024.chinesechess.rule.ChessPosition;

/**
 * 士
 * @author Hu Lang
 */
public class ChessG extends AbstractChess {
    @Override
    public boolean canGoTo(ChessPosition destPos) {
        // 只许沿九宫斜线走单步，可进可退
        return Math.abs(destPos.row - pos.row) == 1 && Math.abs(destPos.col - pos.col) == 1
            && ChessK.isInKingHome(this, destPos);
    }
}

File: client/src/rule/chess/ChessG.ts
import ChessHost from "../chess_host";
import ChessPos from "../ChessPos";
import Game from "../Game";
import AbstractChess from "./AbstractChess";
import { isInKingHome } from "./move_rules";

/**
 * 士
 */
export default class ChessG extends AbstractChess {
    constructor(pos: ChessPos, host: ChessHost) {
        super(pos, host);
    }

    canGoTo(destPos: ChessPos, game: Game) {
        // 只许沿九宫斜线走单步，可进可退
        return Math.abs(destPos.row - this.pos.row) == 1
            && Math.abs(destPos.col - this.pos.col) == 1
            && isInKingHome(this, destPos, game);
    }
}
